만국 박람회  [STEP 1] Taeangel, dudu

FirstDo opened the pull request:
@jungseungyeo 
안녕하세요 린생 😀
만국 박람회 [STEP 1] PR 보냅니다. 잘 부탁드립니다!

## UML
![](https://i.imgur.com/8E70lBc.png)

---

## 고민한점

### parse 매서드의 구현

처음에는 `Heritage`, `Expo` 구조체 각각에 `static parse` 매서드를 구현했는데, 하나의 함수로 만드는게 좋을거 같아서 `ParseManager<T: Decodable>` 구조체를 만들고, `parse`를 구현했습니다.

Decode할 타입이 `[Heritage]`, `Expo` 이기도 하고, 각 타입에 대한 assetName이 달라서 하나의 매서드로 구현할 수 없을 줄 알았는데, `ParseManager<T: [Hertige]>.parse()` 이런식으로 사용할 수 있다는걸 알아서, 하나의 매서드로 합칠 수 있었습니다.

근데 T에는 Decodable을 채택한 타입만 올 수 있는데, `[Heritage]`가 들어갈 수 있더라구요!  
`Heritage` 가 Decodable을 채택했기때문에, `[Heritage]`도 자동으로 Decoable이 채택된 것인지 궁금합니다

### ParseManager 파일의 위치

해당 struct가 Model에 있는게 맞는지 궁금합니다..
Model도 View도 Controller도 아닌것 같은데, 따로 그룹을 만들어줘야 할까요?

FirstDo committed:
add: 필요한 Asset 데이터 추가

expo_assets.zip 파일을 받아서 Xcode Asset에 추가

Taeangel committed:
feat: Model 폴더에 Expo.swift 파일 생성

구조체 Expo에 Decodable 프로토콜 채택

FirstDo committed:
feat: Model 폴더에 Heritage.swift 파일생성

구조체 Heritage에 Codable 채택 및 CodingKeys 구현

Taeangel committed:
feat: Expo와 Heritage 구조체에 parse 메서드 구현, Test코드 구현

dataAssert에서 data를 가져와서 decode한 Model을 반환 하는 함수 구현
제대로 decode가 되는지 확인하는 DecodeTests 구현

FirstDo committed:
style: 주석제거 및 Created by 수정

Taeangel committed:
refactor: 코드베이스로 변경

main.storyBoard 삭제
SceneDelegate와 ViewController에 초기화 코드 작성

FirstDo committed:
refactor: ParseManager 구현

Expo와 Hertitage 구조체에 있는 parse 매서드를 ParseManager의 Generic 타입 매서드로 변경

Taeangel committed:
style: 들여쓰기 수정 및 필요없는 import 제거

FirstDo committed:
style: style convention guide에 맞게 줄바꿈 수정

Taeangel committed:
docs: README 업데이트

FirstDo committed:
style: 컨벤션 수정 및 네이밍 수정

Generic T를 Element, heritages를 decodedData로 수정
컨벤션에 맞게 개행수정

jungseungyeo commented on Expo1900/Expo1900/Model/ParseManager.swift:
```diff
@@ -7,14 +7,14 @@
 
-struct ParseManager<T: Decodable> {
-  static func parse(name: String) -> T? {
+struct ParseManager<Element: Decodable> {
```
element라는 네이밍 너무 좋네요 ㅎㅎ

jungseungyeo commented on Expo1900/Expo1900/Model/Expo.swift:
```diff
@@ -5,7 +5,7 @@
 //
 
-struct Expo: Decodable {
+struct Expo: Codable {
```
Decodable에서 Codable로 변경한 이유가 있나요?

Taeangel replied:
Decodable에서 Codable로 변경한 이유는 UML에 맞추기 위해 Codable로 변경했었고
Codable 보다는 Decodable이 맞는것같아 최종적으로 Decodable로 변경하였습니다.

jungseungyeo commented on Expo1900/Expo1900/AppDelegate.swift:
```diff
@@ -12,16 +12,17 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
   func application(
     _ application: UIApplication,
     configurationForConnecting connectingSceneSession: UISceneSession,
-    options: UIScene.ConnectionOptions) -> UISceneConfiguration {
```
코드를 정리하는 습관 너무 좋습니다. 기본적으로 사전과제에서도 이러한 과정을 해서 제출하면 보다 효과적으로 리뷰하는 사람의 마음을 좋게 할 수 있어요 ㅎㅎ

jungseungyeo commented on Expo1900/DecodeTests/DecodeTests.swift:
```diff
@@ -11,10 +11,10 @@ import XCTest
   
   func test_Expo의_NSDataAsset이_decode_되는지() {
-    XCTAssertNotNil(Expo.parse())
+    XCTAssertNotNil(ParseManager<Expo>.parse(name: "exposition_universelle_1900"))
```
테스트에서는 항상 give, when, then 패턴을 유지해주세요 :)

FirstDo replied:
아래와 같이 변경하였습니다!

jungseungyeo commented on Expo1900/DecodeTests/DecodeTests.swift:
```diff
@@ -11,10 +11,10 @@ import XCTest
 class DecodeTests: XCTestCase {
   
   func test_Expo의_NSDataAsset이_decode_되는지() {
```
한글로 작성하는 네이밍 너무 좋아요! 하지만 몇가지 수정이 되면 좋을거같네요. give, when, then 패턴으로 테스트를 하는 것 처럼 무엇이 준비되면 무엇을 했을때 무슨 결과가된다. 이런식으로 문장으로 끝나고 어떤 결과가 나와야 되는지를 명확히 표현하면 좋을거같아요.

FirstDo replied:
```swift
  func test_assetName이_준비됐을때_Expo타입의_PaseManager_parse의_결과가_Nil이_아니여야한다() {
    //given
    let assetName = "exposition_universelle_1900"
    //when
    let result = ParseManager<Expo>.parse(name: assetName)
    //then
    XCTAssertNotNil(result)
  }
  
  func test_assetName이_준비됐을때_Heritage타입의_PaseManager_parse의_결과가_Nil이_아니여야한다() {
    //given
    let assetName = "items"
    //when
    let result = ParseManager<[Heritage]>.parse(name: assetName)
    //then
    XCTAssertNotNil(result)
  }
```

다음과 같이 변경하였습니다

jungseungyeo replied:
가능하다면 nil을 체크하는것이 아닌 equal로 비교하면 더욱 좋을거같아요 :)

jungseungyeo commented on Expo1900/Expo1900/Model/Heritage.swift:
```diff
@@ -0,0 +1,20 @@
+  let name: String
+  let imageName: String
+  let shortDescription: String
+  let description: String
```
모델링 너무 좋습니다. 하지만 옵셔널이면 안전할거같아요. 서버에서 값이 없어지거나 특정값이 내려오지 않는 경우 값이 매핑이 안되는 이슈가 발생 할 수 있어요.

Taeangel replied:
```swift
struct Heritage: Decodable {
  let name: String?
  let imageName: String?
  let shortDescription: String?
  let description: String?
  
  enum CodingKeys: String, CodingKey {
    case name
    case imageName = "image_name"
    case shortDescription = "short_desc"
    case description = "desc"
  }
}

```

```swift
struct Expo: Decodable {
  let title: String?
  let visitors: Int?
  let location: String?
  let duration: String?
  let description: String?
}
```
이와 같이 옵셔널 타입으로 변경하였습니다.

jungseungyeo reviewed (changes requested):
태엔절, 두두 고생하셨어요!

ParseManager<T: Decodable>를 사용하여 제네릭을 받아서 처리하는 것이 좋았습니다. 또 static 함수를 사용하여 객체를 만들어서 사용하지 않고 바로 사용하는 점도 좋았습니다.

### T에는 Decodable을 채택한 타입만 올 수 있는데, [Heritage]가 들어갈 수 있더라구요!
Heritage 가 Decodable을 채택했기때문에, [Heritage]도 자동으로 Decoable이 채택된 것인지 궁금합니다

`넵 안의 값이 프로토콜을 채택하고 있어 들어갈 수 있습니다. `

### ParseManager 파일의 위치

`흠.. 위치는 참 어려운것이죠. 이것은 주로 사용하는 쪽에서 가까운곳에 있는게 개인적으로는 편한거 같아요. 그래서 현재 Model에서 사용하는 경우가 많아 Model 파일 안에서 따로 Manager라는 폴더를 만들고 사용하거나 Parse와 같은 네이밍으로 만들면 한번에 무슨 역할을 하는 매니져인지 인지가 되어 가독성이 높아 질거 같아요.`

태엔절, 두두 고생하였습니다. 코멘트에서 몇가지 질문과 수정 할 것을 적어두었습니다. 수정 부탁드립니다.
또 하시다가 궁금한 점이 있으면 언제든지 알려주세요 :)

Taeangel committed:
refactor: Expo와 Heritage 프로퍼티를 옵셔널 타입으로 변경, Codable을 Decodable로 변경

값이 존재하지 않을때 fatal오류를 방지하기 위해 옵셔널 타입으로 변경
Encodable을 사용하지 않아 Codable을 Decodable로 변경

FirstDo committed:
test: test함수 네이밍 변경 및 given, when, then 패턴 추가

FirstDo committed:
chore: 잘못입력한 주석 제거

Taeangel committed:
style: Model 그룹에 Manager 그룹생성

ParseManager.swift를 Manager그룹으로 이동

FirstDo committed:
style: ExpoViewController로 네이밍 변경 및 final 키워드 추가

FirstDo committed:
docs: README 업데이트

jungseungyeo commented:
태엔젤, 두두 고생하셨어요. 다음 step 진행하면 좋을거같아요. 테스트에서 nil을 체크하지말고 특정 값을 체크하는 방식으로 다음 스텝에서 해주시면 좋을거같아요. 고생하셨습니다 :)
